File: 1. basic/src/com/hspedu/chapter15/p1/p1.py
from abc import ABC, abstractmethod
from typing import List, TypeVar, Generic

from regex import D

T = TypeVar('T')


class p1(Generic[T]):
    @ staticmethod
    def returnSameType(args: T) -> T:
        return args

    @staticmethod
    def main() -> None:
        animalArr: List[Animal[int]] = []
        # 不谢类型会显示unknown...
        dog: Dog[int] = Dog(0)
        cat: Cat[int] = Cat(1)
        animal: Animal[int] = Animal(2)
        animalArr.append(dog)
        animalArr.append(cat)
        animalArr.append(animal)

        args_: int = cat.getArgs()
        print(args_)  # 1

        res1: str = p1.returnSameType("abc")  # abc
        print(res1)
        res2: List[Animal[int]] = p1.returnSameType(animalArr)
        print(res2)  # [Dog, Cat, Animal]


class IAnimal(Generic[T], ABC):
    @abstractmethod
    def getArgs() -> T: ...


class Animal(Generic[T], IAnimal[T]):
    __args: T

    def __init__(self: 'Animal', args: T) -> None:
        self.__args = args

    def getArgs(self) -> T:
        return self.__args

    def __str__(self) -> str:
        return str(self.__class__.__name__)

    def __repr__(self) -> str:
        return str(self.__class__.__name__)


class Cat(Animal, Generic[T]):
    def __init__(self: 'Cat', args: T) -> None:
        super().__init__(args)


class Dog(Animal, Generic[T]):
    def __init__(self: 'Dog', args: T) -> None:
        super().__init__(args)


p1.main()

File: 1. basic/src/com/hspedu/chapter15/p1/p1.ts
class p1 {
    public static returnSameType<T>(args: T): T {
        return args;
    }
    public static main(): void {
        const animalArr: Animal<number>[] = [];
        // 简单类型可以不传，能推断出来
        const dog = new Dog(0);
        const cat = new Cat(1);
        const animal = new Animal(2);
        animalArr.push(dog, cat, animal);

        const args_: number = cat.getArgs();
        console.log(args_); // 1

        const res1: string = p1.returnSameType('abc');
        console.log(res1); // abc
        const res2: Animal<number>[] = p1.returnSameType(animalArr);
        console.log(res2); // [ Dog { args: 0 }, Cat { args: 1 }, Animal { args: 2 } ]
    }
}

interface IAnimal<T> {
    getArgs(): T;
}
// type也能写成这样，但无法让类实现
type AnimalType<T> = { getArgs: () => T };

class Animal<T> implements IAnimal<T> {
    private args: T;
    public constructor(args: T) {
        this.args = args;
    }
    public getArgs(): T {
        return this.args;
    }

    public toString(): string {
        return this.constructor.name;
    }
}

class Cat<T> extends Animal<T> {
    public constructor(args: T) {
        super(args);
    }
}

class Dog<T> extends Animal<T> {
    public constructor(args: T) {
        super(args);
    }
}

p1.main();
